### ISSUE TITLE
[FEATURE] EncoderFilter

### ISSUE THREAD

**Issue by koaning** [koaning/scikit-fairness#30]
The `InformationFilter` works but it has a downside: it can only filter away linear information. So what if we do a dimensionality reduction technique that might filter away non-linearities? 

Take this neural network: 

![image](https://user-images.githubusercontent.com/1019791/80859723-365df680-8c63-11ea-833d-100532a6866d.png)

By having a network set up this way we can get two gradient signals. One will tune the encoding, the other will filter away information that has predictive power for sensitive attributes. 

![image](https://user-images.githubusercontent.com/1019791/80859746-5e4d5a00-8c63-11ea-9e12-0232d98dc2d3.png)

**Comment by MBrouns** [comment]
I really like this! Some things worth discussing:

- What will we use for the implementation? I'm leaning to `jax` atm because it's not a giant dependency like `torch`. We can then also replace the current `autograd` code with `jax`
- Will the transform return the bottleneck layer or the final reconstructed layer? I guess the reconstructed one as that will reflect the original features as much as possible
- Do we train the adverserial on the bottleneck or the final reconstructed layer? I have a feeling that if you train it on the bottleneck the decoder might add some of the unfairness back into the dataset.

**Comment by koaning** [comment]
Now that I think about your comments I feel the need to redraw. 

This was my original idea, and it comes with a weakness. 

![image](https://user-images.githubusercontent.com/1019791/80910645-46440c00-8d31-11ea-9a46-472de14dc082.png)

We're trying to reconstruct race in two places, this would be better; 

![image](https://user-images.githubusercontent.com/1019791/80910665-6673cb00-8d31-11ea-998e-3933eac8e4ce.png)

In the latter example I'd be fine with giving the option to the user what to grab; the bottleneck or the output layer. 

# Expand the idea 

If I expand on the diagrams then I think we may even be able to give the user two variants.

## Variant One 

Note that the encoder no longer outputs the sensitive attribute in this diagram. 

![image](https://user-images.githubusercontent.com/1019791/80910846-a7b8aa80-8d32-11ea-9db2-eaf33600ef38.png)

## Variant Two

Also here; the encoder no longer outputs the sensitive attribute in this diagram. 

![image](https://user-images.githubusercontent.com/1019791/80910857-b3a46c80-8d32-11ea-8064-410b12417b2d.png)

@MBrouns would you agree that it would be best to these two variants as two different preprocessors, both of which allow the user to select if they want the bottleneck or the encoder output? There is a subtle change in the dynamics of the gradient ...

**Comment by koaning** [comment]
@MBrouns and yes, it seems to be my preference to do this in jax. I fear it is a little bit experimental, but I prefer it because it is more lightweight.

**Comment by koaning** [comment]
Mhmm ... I'm running basic examples locally ... this stuff is slow. Even on the boston dataset.

**Comment by koaning** [comment]
Aaand this is adverserial learning, so the loss graphs get freaky. 

![image](https://user-images.githubusercontent.com/1019791/81344725-a3381d00-90b7-11ea-8264-630cecaad228.png)
